### ISSUE TITLE
Direct CLI caps BLAS threads to 2 process-wide; service does not (the 1.30x attribution is REFUTED — see comment)

### ISSUE THREAD

**Issue by pcalnon** [pcalnon/juniper-cascor#531]
## Summary

The direct CLI and the service tier load BLAS with **different thread-pool sizes**, and the difference is not configuration — it is which entry-point file the process happened to execute. Measured effect: the direct CLI's candidate phase runs **1.52×** the service's on identical data, identical initialisation and an identical config, and **1.30× of that is attributable to this asymmetry alone**.

`src/main.py:48-50` (the **direct-CLI** entry point):

```python
os.environ.setdefault("OMP_NUM_THREADS", "2")
os.environ.setdefault("MKL_NUM_THREADS", "2")
os.environ.setdefault("OPENBLAS_NUM_THREADS", "2")
```

The **service** enters through `uvicorn api.app:create_app --factory` and never executes `main.py`. `git grep -nE "NUM_THREADS|set_num_threads" -- src/api` returns **nothing**. So the service loads BLAS with these unset.

Why that is permanent rather than cosmetic, in `main.py`'s own words (`:41-45`): the variables *"are read once when the BLAS library is first loaded; setting them after import has no effect."* Candidate workers are created from a **forkserver** context, so each worker inherits the BLAS pool of the process it descends from and can never resize it. `train_candidate_worker` re-exports the same three variables (`cascade_correlation.py:3874-3876`) but that is inert for exactly the reason the comment gives — only its adjacent `torch.set_num_threads()` call has any effect.

## Measurements

Probe suite `e-k-thread-probe-cap16`, cap 16 / pool 8, `dataset.params.seed: 42` so **both arms share the dataset AND the network initialisation** (the service network is unconditionally seed 42 — see #530). All four arms cap-bound at 16 iterations. Only the inherited thread budget varies; `main.py` uses `setdefault`, so an exported value wins.

| arm | candidate phase | candidate epochs | s / candidate epoch | total span |
| --- | --- | --- | --- | --- |
| **service** (unset) | 944 s | 44,910 | 0.02102 | **961 s** |
| CLI, `OMP=16` | 1103 s (1.17×) | 46,080 (1.03×) | 0.02394 (1.14×) | 1156 s |
| CLI, `OMP=8` | 1240 s (1.31×) | 51,630 (1.15×) | 0.02402 (1.14×) | 1293 s |
| CLI, **default → 2** | 1432 s (1.52×) | 54,220 (1.21×) | 0.02641 (1.26×) | 1485 s |

The **output** phase is 16 s on all four arms — this is confined to candidate training, which is ~98% of the span at these caps.

Two distinct effects, both from the same cause:

1. **Throughput** — 1.26× → 1.14× as the budget rises (plateauing between 8 and 16). Expected.
2. **Epoch count** — 1.21× → **1.03×**. Less obvious and arguably more important: thread count changes BLAS reduction order, hence floating-point results, hence where a patience-based candidate early-stopping loop terminates. At a matched budget the CLI runs essentially the *same* number of candidate epochs as the service; at 2 threads it runs 21% more. So the cap does not merely slow each epoch, it causes **more epochs to be run**.

**Residual: 1.17× remains at matched thread count.** This issue does not claim to explain that part; it is being investigated separately.

## Why this is a defect and not a tuning preference

RC-1 (commit `aa46ad5`, "parallelize candidate training by resolving 4 root causes of serial execution") introduced these lines. Its Phase 1 is described as *"Pin PyTorch threads to 1 per worker, set BLAS env vars before library load"*, and the stated target is *"PyTorch thread pool oversubscription (N workers × M threads per worker)"*. The in-file comment says the purpose is *"to prevent the **parent process** from spawning excessive internal threads that would compete with worker processes"*.

Three observations:

- **The parent is already limited, separately and correctly.** `_init_multiprocessing` sets `torch.set_num_threads(max(2, worker_thread_count * 2))` (`cascade_correlation.py:1126`). The env-var mechanism is not what protects the parent.
- **The mechanism is broader than the intent.** It targets the parent but, being process-wide and inherited across fork, it permanently caps every candidate **worker** — which is where the measured cost lands.
- **The service is a live control.** It has never had these variables, runs the candidate pool fully parallel (8 forked workers at ~90% CPU each, confirmed by `ps` during a run), and is the *faster* of the two paths. So the anti-oversubscription protection does not depend on the env vars; the per-worker `torch.set_num_threads(max(1, worker_thread_count))` (`:3873`, default `worker_thread_count = 1`) is doing that job on both paths.

## Proposed fix

The specific bug is that a *policy* lives in one of two entry points. Whatever the policy should be, both entry points must apply the same one.

1. Add a single shared helper — e.g. `cascor_constants` or a small `runtime_threads` module — exposing `configure_blas_threads()`, and call it from **both** `main.py` (before any BLAS-importing module) and the service app factory.
2. **Default to not setting the variables**, matching today's service behaviour, which is the empirically faster of the two and the one every service-tier measurement to date was taken under. This is the behaviour-preserving choice for the service and a fix for the CLI.
3. Make it overridable by an explicit env knob (e.g. `JUNIPER_CASCOR_BLAS_THREADS`) for anyone who does want to cap it, so the capability RC-1 added is retained rather than deleted.
4. Keep `worker_thread_count` / `torch.set_num_threads()` exactly as they are — that is RC-1's actual oversubscription fix and it is not in question.

Worth doing in the same change: drop or comment the inert `os.environ[...] = ...` re-export in `train_candidate_worker` (`:3874-3876`), since it cannot work after BLAS load and currently reads as though the worker controls its own pool.

## Reproduction

```bash
# service reference
python util/experiments/run_suite.py --suite util/experiments/suites/p4/e-k-thread-probe-cap16.yaml

# the same generated cell config through the CLI at 2 / 8 / 16 threads
util/ad-hoc/2026-08-17_h2h_thread_campaign.bash <SUITE_DIR> <OUT_ROOT> <CASCOR_SRC>

# decomposition
python util/ad-hoc/2026-08-16_h2h_phase_split.py <arm>/logs ...
grep -c "candidate_unit.py: 695" <arm>/direct_cli.log     # 1 record per 10 candidate epochs
```

(Tooling in juniper-ml; the wide-budget head-to-head that surfaced this is juniper-ml#1143.)

**Comment by pcalnon** [comment]
## Correction: the 1.30× attribution does not reproduce

This issue's title and body attribute **1.30× of a 1.52×** candidate-phase penalty to `main.py`'s `OMP=2` cap. That figure came from **one run per setting**. Re-measured 2026-08-22, rep-paired at k=3 with conditions rotated *inside* each replicate (so host drift hits all conditions equally):

| rep-paired rate ratio vs `unset` | mean ± sd | 95% CI | |
| --- | ---: | --- | --- |
| explicit `16` | 1.033 ± 0.095 | [0.926, 1.141] | includes 1.0 |
| **`2`** (this cap) | **1.016 ± 0.116** | **[0.885, 1.148]** | includes 1.0 |

Candidate work was also unchanged (49,460 vs 49,550 epochs), so the combined phase effect is ~**1.018×**. The interval's upper bound is 1.148, so **1.30× is excluded**. n=3 cannot rule out a modest ~10% effect, but the headline attribution does not survive.

Corroborating: **#533 did not measurably move the cap-64 gap** — 1.924 ± 0.486 post-#533 against a pre-#533 1.99 ± 0.21, overlapping. There was little there to remove.

**#533 remains the right change.** One BLAS policy called from both entry points is correct engineering whatever the number says, and the asymmetry it fixed was real. What does not survive is its *performance* justification.

**The real cause of the CLI-vs-service penalty has since been found and fixed** ([#563](https://github.com/pcalnon/juniper-cascor/pull/563)): `inspect.getmodule`, reached from the logger's per-record caller resolution, scans `sys.modules` on every call — ~78% of candidate-worker CPU, and O(len(`sys.modules`)), which is why it hit the CLI (1,871 modules) harder than the service (1,410). Fixing it made training ~9× faster on **both** arms and removed the per-epoch penalty entirely.

Evidence: `notes/JUNIPER_2026-08-21_JUNIPER-ECOSYSTEM_CLI-EXPERIMENTATION-RESIDUAL-WALL-GAP-EVIDENCE.md` §4.3a (juniper-ml).
